cpp/src/arrow/types/primitive.cc

```diff
@@ -41,12 +42,25 @@ bool PrimitiveArray::EqualsExact(const PrimitiveArray& other) const {
     return false;
   }
 
-  bool equal_data = data_->Equals(*other.data_, length_);
   if (null_count_ > 0) {
-    return equal_data &&
-      nulls_->Equals(*other.nulls_, util::ceil_byte(length_) / 8);
+    bool equal_bitmap = nulls_->Equals(*other.nulls_, util::ceil_byte(length_) / 8);
+    if (!equal_bitmap) {
+      return false;
+    }
+
+    const uint8_t* this_data = raw_data_;
+    const uint8_t* other_data = other.raw_data_;
+
+    for (int i = 0; i < length_; ++i) {
+      if (!IsNull(i) && memcmp(this_data, other_data, value_size_)) {
```

Comment:
@danrobinson here you see the problem with uninitialized slots that you brought up on the ML in action. This can obviously can (and should) be special-cased for 1, 2, 4, and 8 bytes using integer comparisons. I am not sure it is going to be safe in general to assume that Arrow writers have zero'd out the uninitialized slots =|